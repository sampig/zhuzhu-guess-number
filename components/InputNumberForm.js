import React, { useCallback, useRef, useState } from 'react'
import { Button, StyleSheet, TextInput, View } from 'react-native'
import i18n from 'util/i18n-utils'

const styles = StyleSheet.create({
  container: {
    flexDirection: 'row',
    alignItems: 'center',
    backgroundColor: '#ffffff',
    paddingRight: 6,
  },
  textInput: {
    flex: 1,
    borderWidth: 2,
    fontSize: 20,
    margin: 6,
    borderRadius: 5,
    padding: 6,
  },
})

/**
 * Component to display a form of input a guess.
 * @param {object} gameParam  - the gameParam object including game, guessList and functions
 * @param {function} tryGuess - a function to guess
 * @returns {node}            - React node
 * @constructor
 */
function InputNumberForm ({ gameParam, tryGuess }) {
  const [value, setValue] = useState('')
  const inputRef = useRef()

  const submitGuess = useCallback(() => {
    inputRef?.current.blur()
    tryGuess(value.split(''))
    setValue('')
  }, [inputRef, tryGuess, value, setValue])

  /**
   * Check if the game is playing.
   * @returns {boolean} - true if it is started
   */
  const isGameStarted = () => {
    return !(!gameParam?.game?.number || gameParam?.isWin() || gameParam?.isGameOver())
  }

  /**
   * Check if 4 digits are input
   * @returns {boolean} - true if 4 digits are input
   */
  const isInputValid = () => {
    if (!value || value.split('').length !== 4) {
      return false
    } else {
      return value.split('').findIndex(n => n === '' || isNaN(Number(n)) || Number(n) < 0 || Number(n) > 9) < 0
    }
  }

  return (
    <View style={styles.container}>
      <TextInput
        ref={inputRef}
        style={styles.textInput}
        value={value}
        onChangeText={text => setValue(text)}
        autoCapitalize='none'
        editable={isGameStarted()}
        placeholder={i18n.t('enterNumber')}
      />
      <Button
        title={i18n.t('guess')}
        disabled={!isGameStarted() || !isInputValid()}
        onPress={submitGuess}
      />
    </View>
  )
}

export default InputNumberForm
