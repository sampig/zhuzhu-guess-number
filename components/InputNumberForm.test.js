import React from 'react'
import renderer from 'react-test-renderer'
import InputNumberForm from './InputNumberForm'

describe('The InputNumberForm component', () => {
  it('renders without crashing', () => {
    const render = renderer.create(<InputNumberForm tryGuess={jest.fn()} />).toJSON()
    expect(render).toMatchSnapshot()
  })
})
