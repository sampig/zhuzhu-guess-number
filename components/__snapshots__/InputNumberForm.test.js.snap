// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The InputNumberForm component renders without crashing 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#ffffff",
      "flexDirection": "row",
      "paddingRight": 6,
    }
  }
>
  <TextInput
    allowFontScaling={true}
    autoCapitalize="none"
    editable={false}
    onChangeText={[Function]}
    placeholder="Enter a 4-digit number"
    rejectResponderTermination={true}
    style={
      Object {
        "borderRadius": 5,
        "borderWidth": 2,
        "flex": 1,
        "fontSize": 20,
        "margin": 6,
        "padding": 6,
      }
    }
    underlineColorAndroid="transparent"
    value=""
  />
  <View
    accessibilityRole="button"
    accessibilityState={
      Object {
        "disabled": true,
      }
    }
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {},
          Object {},
        ]
      }
    >
      <Text
        disabled={true}
        style={
          Array [
            Object {
              "color": "#007AFF",
              "fontSize": 18,
              "margin": 8,
              "textAlign": "center",
            },
            Object {
              "color": "#cdcdcd",
            },
          ]
        }
      >
        Guess
      </Text>
    </View>
  </View>
</View>
`;
